### PR TITLE
Synchronizing access to register remote query handlers

### PR DESCRIPTION
**Pull request by oxtoacart**
For getlantern/lantern-internal#2767

### CODE REVIEW

File: table.go [hunk 1]
```diff
@@ -264,12 +264,15 @@ func addPointsField(fields core.Fields) core.Fields {
 }
 
 func (t *table) startFollowing(walOffset wal.Offset) {
+	t.db.followMx.Lock()
 	newSubscriber := t.db.newStreamSubscriber[t.From]
 	if newSubscriber == nil {
 		newSubscriber = make(chan *tableWithOffset, 100)
+		log.Debugf("Following leader for %v", t.From)
 		go t.db.followLeader(t.From, newSubscriber)
```
Comment (oxtoacart) [inline table.go]: This is the actual fix. It used to be possible for us to spawn multiple `followLeader` calls for the same stream. The mutex prevents that.

File: zenodb_test.go [hunk 2]
```diff
@@ -103,9 +105,9 @@ func doTestCluster(t *testing.T, numPartitions int, partitionBy []string) {
 					var register func()
 					register = func() {
 						leader.RegisterQueryHandler(partition, func(ctx context.Context, sqlString string, isSubQuery bool, subQueryResults [][]interface{}, unflat bool, onFields core.OnFields, onRow core.OnRow, onFlatRow core.OnFlatRow) (interface{}, error) {
-							// Re-register when finished
-							defer register()
-							return query(ctx, sqlString, isSubQuery, subQueryResults, unflat, onFields, func(key bytemap.ByteMap, vals core.Vals) (bool, error) {
+							// Re-register immediately
+							go register()
```
Comment (oxtoacart) [inline zenodb_test.go]: This was needed to make test work now that we're locking.

**Comment by oxtoacart** [comment]
The CI failure is not new. I've logged a separate ticket for that. I think we should get this in and deployed to fix the restart issue.

File: zenodb.go [hunk 3]
```diff
@@ -153,6 +153,7 @@ type DB struct {
 	requestedIterations  chan *iteration
 	coalescedIterations  chan []*iteration
 	closed               bool
+	followMx             sync.Mutex
```
Comment (hwh33) [inline zenodb.go]: This protects the newStreamSubscriber map, right?  It's not obvious from the name.  It's easy to figure out if you see this mutex and wonder what it's for.  However, I could also see someone down the road modifying the newStreamSubscriber map without realizing there's an associated mutex they should be using.

**Comment by hwh33** [comment]
LGTM!

**Comment by oxtoacart** [comment]
Thanks @hwh33!

**Pull request merged**